### ISSUE TITLE
[Question] Correct way to generate sfm_data with multiple cameras

### ISSUE THREAD

**Issue by park-sungjoon** [openMVG/openMVG#2196]
**Problem**
I am trying to reconstruct dense point cloud using images taken from 8 synchronized cameras. To maximize the accuracy, I found the intrinsic parameters for each of the cameras. However, I am not sure what is the correct way to use these intrinsic parameters in the openMVG and openMVS pipeline.

**Details on what I have tried**
I am basically using `SfM_SequentialPipeline.py` in openMVG repository to generate sparse point cloud, which is then exported for use in openMVS with **openMVG_main_openMVG2openMVS**. I then use **DensifyPointCloud** to generate dense point cloud.

I have no problem generating dense point cloud when using the same camera intrinsics for all of the images (I called **openMVG_main_ConvertList** with option "-k f;0;ox;0,f;oy;0;0;1"). Of course I could just ignore slight differences in the camera intrinsic parameters. But if I were using cameras with very different intrinsics, this would not work, and I was wondering what is the correct way to approach this problem.

What I have tried is as follows: 
In the documentation for image listing [(SfMInit_ImageListing),](https://openmvg.readthedocs.io/en/latest/software/SfM/SfMInit_ImageListing/) it is noted that for old version of openMVG, one can use **openMVG_main_ConvertList** to generate sfm_data.json, which is nice because I can insert the intrinsic parameters corresponding to each of the images.  Although I am using the latest branch for both openMVG, I noticed that this function is still available, so I tweaked the `SfM_SequentialPipeline.py` by using **openMVG_main_ConvertList** instead of **openMVG_main_SfMInit_ImageListing**. This seems to work without problem (sparse point cloud generated by openMVG looks good) until I call **DensifyPointCloud**, which generates errors like 
"error: opening file '/dataset/000000/scene_undistorted_images/depth0000.dmap' for reading depth-data"
Generated sfm_data.json: [sfm_data.zip](https://github.com/openMVG/openMVG/files/11476451/sfm_data.zip)
Full log for DensifyPointCloud: [DensifyPointCloud.log](https://github.com/openMVG/openMVG/files/11476465/DensifyPointCloud.log)

I also tried manually inserting intrinsics in the sfm_data.json (I inserted the intrinsics data, and changed the "id_intrinsic" in the "views" to point to the intrinsics data)

What is the correct way to list the camera intrinsics?

**Comment by park-sungjoon** [comment]
@zhaozhongch  Yes, I think it's basically the same issue. Actually, I asked this question [here](https://github.com/cdcseacave/openMVS/issues/989)
According to cdcseacave,
> Once that is done, you do not have to do anything special in OpenMVS, it will simply use the calibrated list of cameras from the received MVS file.

So I thought that maybe the problem is with the way openMVG is generating sfm_data.bin and/or scene.mvs

This is not directly related to this issue, but when I added -e NONE option to openMVG_main_SfM, camera extrinsics were still getting modified for some reason, and I had to use openMVG_main_ComputeStructureFromKnownPoses instead

**Comment by zhaozhongch** [comment]
Hi thanks for the discussion.
You mean you tried to use 
openMVG_main_ComputeStructureFromKnownPoses with sfm_data.json as input then it generates a sfm_data.bin? Does it work?
I checked the [tutorial code](https://github.com/openMVG/openMVG/blob/b9024c3117f4df283bb58c6474ee2c224e6c884e/src/software/SfM/tutorial_demo.py.in#L74)  seems it uses a sfm_data.bin as input instead of json file.

**Comment by park-sungjoon** [comment]
I wasn't aware of the tutorial code, but yes, you can use sfm_data.json as input. I'm basically calling
```
openMVG_main_SfMInit_ImageListing -i images/ -o matches/ -k "f;0;ox;0;f;oy;0;0;1"
openMVG_main_ComputeFeatures -i matches/sfm_data.json -o matches/ -m SIFT -p ULTRA
openMVG_main_PairGenerator -i matches/sfm_data.json -o matches/pairs.bin -m 0
openMVG_main_ComputeMatches -i matches/sfm_data.json -p matches/pairs.bin -o matches/matches.putative.bin
openMVG_main_GeometricFilter -i matches/sfm_data.json -m matches/matches.putative.bin -g f -o matches/matches.e.bin
openMVG_main_ComputeStructureFromKnownPoses -i  matches/sfm_data.json -m  matches/ -o reconstruction_global/sfm_data.bin -f matches/matches.e.bin -d
openMVG_main_openMVG2openMVS -i reconstruction_global/sfm_data.bin -o reconstruction_global/scene.mvs -d scene_undistorted_images
DensifyPointCloud -i reconstruction_global/scene.mvs -w scene_undistorted_images/
```
Regardless of whether I use openMVG_main_SfM or openMVG_main_ComputeStructureFromKnownPoses, DensifyPointCloud didn't work when I try to specify camera intrinsics for each of the cameras.

**Comment by HenghengHe** [comment]
I have the same questions. Did you solve it?

**Comment by pmoulon** [comment]
OpenMVG can handle intrinsic group so multiple camera can share the same intrinsics during the optimization. See the sfm_data json file. There is no support for shared relative pose (it is supported in a branch but not well documented).
